kubespawner/spawner.py

```diff
@@ -11,45 +11,61 @@
 import threading
 import sys
 from urllib.parse import urlparse, urlunparse
-from concurrent.futures import ThreadPoolExecutor
+import json
 import multiprocessing
-
+from concurrent.futures import ThreadPoolExecutor
 
 from tornado import gen
 from tornado.ioloop import IOLoop
 from tornado.concurrent import run_on_executor
-from traitlets.config import SingletonConfigurable
 from traitlets import Type, Unicode, List, Integer, Union, Dict, Bool, Any
 from jupyterhub.spawner import Spawner
+from jupyterhub.utils import exponential_backoff
 from jupyterhub.traitlets import Command
 from kubernetes.client.rest import ApiException
 from kubernetes import client
 import escapism
 
 from kubespawner.traitlets import Callable
-from kubespawner.utils import request_maker, k8s_url
 from kubespawner.objects import make_pod, make_pvc
-from kubespawner.reflector import PodReflector
+from kubespawner.reflector import NamespacedResourceReflector
 
 
-class SingletonExecutor(SingletonConfigurable, ThreadPoolExecutor):
-    """
-    Simple wrapper to ThreadPoolExecutor that is also a singleton.
+class PodReflector(NamespacedResourceReflector):
+    kind = 'pods'
 
-    We want one ThreadPool that is used by all the spawners, rather
-    than one ThreadPool per spawner!
-    """
-    pass
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+    }
+
+    list_method_name = 'list_namespaced_pod'
+
+    @property
+    def pods(self):
+        return self.resources
 
 class KubeSpawner(Spawner):
     """
     Implement a JupyterHub spawner to spawn pods in a Kubernetes Cluster.
     """
+
+    # We want to have one threadpool executor that is shared across all spawner objects
+    # This is initialized by the first spawner that is created
+    executor = None
+
+    # We also want only one pod reflector per application
+    pod_reflector = None
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # By now, all the traitlets have been set, so we can use them to compute
         # other attributes
-        self.executor = SingletonExecutor.instance(max_workers=self.k8s_api_threadpool_workers)
+        if self.__class__.executor is None:
+            self.__class__.executor = ThreadPoolExecutor(
+                max_workers=self.k8s_api_threadpool_workers
+            )
+        self.executor = self.__class__.executor
```

Comment:
This extra assignment is not necessary. `self.executor` resolves to the class attribute already.